### ISSUE TITLE
Catadioptric system

### ISSUE THREAD

**Issue by CAClaveau** [singer-yang/DeepLens#11]
Hello,

First of all, thank you very much for sharing your amazing tool!

I’ve only quickly checked the repo and your papers, so apologies if I missed the information. I noticed there is a Class Surface « Mirror ».

Can DeepLens model a reflective system (for example, a telescope)? Do you have such an example? Even if it might be somewhat different from the primary application (computational imaging) for which DeepLens was initially designed.

Thank you for your consideration!

**Comment by singer-yang** [comment]
Hello CAClaveau,

Thank you for your interest! While we do have optical systems that utilize mirror surfaces and other reflective elements, we currently have no plans to release them in the near future.

You can create your own optical systems using the DeepLens framework and its functions. I can also provide some simple examples involving mirror surfaces when I have time.

Best,  
Xinge

**Comment by CAClaveau** [comment]
Hi @singer-yang,

Thank you for your prompt reply and clarifications!

Your examples showing how to load a predefined optical configuration from a JSON file via the `GeoLens` class (for a sequential, rotationally symmetric refractive optical system, correct?) are already very helpful in most cases.

Yes, having a few examples on how to directly build a custom optical system, including reflective surfaces/mirrors, could be useful at some point. Please let me know when such examples are added, if that happens. In the meantime, I’ll have some time to familiarize myself further with DeepLens and the code.

Thanks again for your consideration.

Best,
CAClaveau

**Comment by CAClaveau** [comment]
(btw, I believe that for some of the example scripts provided, the `GeoLens `Class no longer accepts the "sensor_res" parameter during instantiation since the latest published version of DeepLens - unless I'm mistaken.)

**Comment by singer-yang** [comment]
Hello CAClaveau,

Thank you again for your interest! In the latest version, we've updated the lens initialization code. Specifically, "sensor_res" has been moved to the GeoLens.read_lens_json() function and needs to be specified in the .json file. Additionally, we are currently using "roc" for GeoLens IO to align with Zemax settings for better clarity.

To build a coaxial optical system with mirrors, you can modify the GeoLens.trace() function to define the ray tracing sequence. For non-coaxial systems, it’s a bit more complex, but I will do my best to provide example support code. :-)

Best,
Xinge

**Comment by CAClaveau** [comment]
Hi @singer-yang,

Thank you for the additional information - it’s helpful!

Got it, I’ll take a look at the `GeoLens.trace` function.

Sounds good, thanks! Please let me know when you add examples of non-coaxial/non-sequential optical systems, including reflective surfaces, if you have time to work on that.

Best,
CAClaveau

**Comment by CAClaveau** [comment]
Hi @singer-yang 

I’m still getting familiar with the code - it’s quite fun experimenting with the examples!

If it may be useful, I’ve noticed the following:

1. In the function `GeoLens.calc_foc_dist` at line #1616, I think the function `math.mean` doesn’t exist (it may be replaced with the mean function from either Torch or NumPy).

2. In the function `GeoLens.read_lens_json` at line #3377, `self.d_sensor = torch.tensor(d)` seems to override the value for "d_sensor" provided in the JSON config file by the user. (so, I’ve been using the very handy function `refocus` to adjust the sensor position to where the image plane forms for a given, fixed object plane.)

More importantly, I tried running the script "5_pupil_field" with the "ef35mm_f2.0.json" system in the "2f-2f" configuration (mag = -1) as a sanity check. I don’t understand why the resulting PSF or diffraction pattern (="psf_coherent") looks like it’s coming from a square aperture rather than a circular one, as shown below (in log scale):

![psf_coherent](https://github.com/user-attachments/assets/1810145e-7051-4fd8-8e22-c64b283021e7)

It’s possible I missed something or need to adjust some parameters. If you could reproduce the example, I’d be very curious to know if you get the same result.

Thanks for your time!
CAClaveau

**Comment by singer-yang** [comment]
Hello CAClaveau,

Thank you for your feedback!

1. I've updated math.mean() to np.mean() and will include this in the next release.
2. When loading a lens from a .json file, I try to be consistent to ZEMAX settings, which only consider the distance from the current surface to the next. Therefore, d_sensor is calculated based on the distance from the last lens surface to the image plane, rather than the value in the .json file. I include d_sensor in the .json file to provide a rough estimate of the camera system's total thickness, although it's not used. I will add a parenthesis to d_sensor in the next release.
3. The diffraction pattern results from the default square shape of the DOE. You can modify the code to support a circular DOE.

Thank you for your interest in our work! Your questions indicate a strong understanding of the code. If you encounter any bugs or wish to add features, please don't hesitate to submit pull requests. :-)

Best,  
Xinge

**Comment by CAClaveau** [comment]
Hi @singer-yang,

Thank you again for your quick reply and the detailed information provided.

Understood regarding the new definition of the "d_sensor" parameter, now consistent with Zemax. ((unless I’m mistaken, I still have the impression - e.g., after calling the function `refocus`, that within the code, "d_sensor" still appears to represent an absolute distance rather than a relative one defined wrt the last surface of the optical system.))

3. FYI, in the script "5_pupil_field" mentioned above, the modeled system is a GeoLens type. There is no DOE, and the aperture is circular. I was finally able to achieve the expected diffraction pattern (~Airy disk) by: 1) increasing the detector size to retain the imprint of the circular aperture in the wavefront calculated by the function `pupil_field`, and 2) reducing the aperture size to ensure sufficient sampling of the wavefront using the "coherent ray tracer" engine - for a fixed maximum resolution "sensor_res".

Thank you again for your consideration. Looking forward to the upcoming developments of DeepLens!

Best,
Charles-Antoine

**Comment by singer-yang** [comment]
Hello Charles,

> after calling the function refocus, that within the code, "d_sensor" still appears to represent an absolute distance rather than a relative one defined wrt the last surface of the optical system

Yes, in the internal functions, using the absolute value of "d_sensor" is more straightforward than using relative distance. However, for lens data I/O, relative distance is easier to understand. This is based on my understanding, and I would be happy to hear any better implementation methods.

> FYI, in the script "5_pupil_field" mentioned above, the modeled system is a GeoLens type. There is no DOE, and the aperture is circular. 

I apologize for my mistake. Now I think the cross-line diffraction pattern might be caused by the square shape of the pupil wavefield. When a square wavefield is propagated without sufficient padding, edge diffraction patterns may appear.

> I was finally able to achieve the expected diffraction pattern (~Airy disk) by: 1) increasing the detector size to retain the imprint of the circular aperture in the wavefront calculated by the function pupil_field, and 2) reducing the aperture size to ensure sufficient sampling of the wavefront using the "coherent ray tracer" engine - for a fixed maximum resolution "sensor_res".

Yes, those are also suggestions I would make. Generally, the more we sample, the better we can get.

Thank you very much for your encouragement! We will continue working on building DeepLens to see how far we can go!  :-)

Best,
Xinge
